### ISSUE TITLE
Deployment to Nexus from M2Eclipse is broken because of parallel upload

### ISSUE THREAD

**Issue by chadj2** [eclipse-m2e/m2e-core#1042]
Numerous posts on message boards are reporting that they get an error 401 when deploying to Nexus and they say that the solution is to use a standalone maven install. (see [this post](https://stackoverflow.com/questions/24830610/why-am-i-getting-a-401-unauthorized-error-in-maven/74318377#74318377)

I think the root cause is that M2Eclipse tries to upload artifacts in parallel. Below you can see the output of a deploy using embedded maven M2elipse. The Jar and POM upload in parallel. The pom succeeds but the jar fails to upload and it returns code "401 Unauthorized". When I look in Nexus I can see the pom but not the jar.

```
[INFO] --- maven-deploy-plugin:3.0.0:deploy (default-deploy) @ xxx-project ---
[INFO] Uploading to : https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.jar
[INFO] Uploading to : https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.pom
[INFO] Uploaded to : https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.pom (13 kB at 36 kB/s)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.527 s
[INFO] Finished at: 2022-11-04T08:38:47-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:3.0.0:deploy (default-deploy) on project ***-project: Failed to deploy artifacts: Could not transfer artifact com.***:***-project:jar:7.1.8.2 from/to ***-nexus (https://nexus.***.com/repository/releases/): Access denied to https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.jar. Error code 401, Unauthorized -> [Help 1]
```

The log below shows a successful deploy made from a standalone maven.

```
[INFO] --- maven-deploy-plugin:3.0.0:deploy (default-deploy) @ ***-project ---
Uploading to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.pom
Uploaded to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.pom (6.4 kB at 12 kB/s)
Uploading to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.jar
Uploaded to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2.jar (22 kB at 77 kB/s)
Uploading to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2-tests.jar
Uploaded to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/7.1.8.2/***-project-7.1.8.2-tests.jar (8.0 kB at 29 kB/s)
Downloading from ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/maven-metadata.xml
Downloaded from ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/maven-metadata.xml (371 B at 3.1 kB/s)
Uploading to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/maven-metadata.xml
Uploaded to ***-nexus: https://nexus.***.com/repository/releases/com/***/***-project/maven-metadata.xml (371 B at 1.4 kB/s)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.860 s
[INFO] Finished at: 2022-11-04T08:54:29-05:00
[INFO] ------------------------------------------------------------------------
```

In summary it seems that Nexus does not like parallel deployment and if you attempt this it will return error 401.

**Comment by mickaelistria** [comment]
Was this issue reported to Nexus?

**Comment by chadj2** [comment]
I don't know if nexus is aware of it and if this is a feature or bug. In any case m2e should work the same way as a standalone Maven install.

**Comment by HannesWell** [comment]
> In any case m2e should work the same way as a standalone Maven install.

That's right, M2E should not behave differently. But 'm not aware of any specialty in M2E in this regard, but also have never done that, so maybe we have an easter-egg here.
The question is why M2E tries to upload in parallel?

For clarification, your start a build with goal `deploy` and your project uses the `nexus-staging-maven-plugin`?
IIRC that has an option to delay artifact upload to the very last module and then upload all at once. Is that the property you referring to as parallel-upload?

**Comment by chadj2** [comment]
> For clarification, your start a build with goal deploy and your project uses the nexus-staging-maven-plugin?

I am using the `deploy` goal with the `maven-deploy-plugin`. The `nexus-staging-maven-plugin` is not compatible with Nexus 3 and the `nxrm3-maven-plugin` only works with the licensed version of Nexus 3.

I think M2E uses a different driver for upload than the standalone maven. I recently figured out that the `-Daether.connector.basic.threads=1` VM option is a valid workaround but that should be the default behavior.

> IIRC that has an option to delay artifact upload to the very last module and then upload all at once. Is that the property you referring to as parallel-upload?

This is a simple single project deployment and so the `deployAtEnd` option does not apply in this case.
